core/src/test/java/org/apache/calcite/test/RelMetadataTest.java

```diff
@@ -3132,6 +3133,35 @@ private void checkNodeTypeCount(String sql, Map<Class<? extends RelNode>, Intege
         is("=($0, $1)"));
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-4315">[CALCITE-4315]
+   * NPE in RelMdUtil#checkInputForCollationAndLimit</a>. */
+  @Test void testCheckInputForCollationAndLimit() {
+    final Project rel = (Project) convertSql("select * from emp, dept");
+    final Join join = (Join) rel.getInput();
+    final RelOptTable empTable = join.getInput(0).getTable();
+    final RelOptTable deptTable = join.getInput(1).getTable();
+    Frameworks.withPlanner((cluster, relOptSchema, rootSchema) -> {
+      checkInputForCollationAndLimit(cluster, empTable, deptTable);
+      return null;
+    });
+  }
+
+  private void checkInputForCollationAndLimit(RelOptCluster cluster, RelOptTable empTable,
+      RelOptTable deptTable) {
+    final RexBuilder rexBuilder = cluster.getRexBuilder();
+    final RelMetadataQuery mq = cluster.getMetadataQuery();
+    final List<RelHint> hints = ImmutableList.of();
+    final LogicalTableScan empScan = LogicalTableScan.create(cluster, empTable, hints);
+    final LogicalTableScan deptScan = LogicalTableScan.create(cluster, deptTable, hints);
+    final LogicalJoin join =
+        LogicalJoin.create(empScan, deptScan, ImmutableList.of(),
+            rexBuilder.makeLiteral(true), ImmutableSet.of(), JoinRelType.INNER);
+    assertTrue(
+        RelMdUtil.checkInputForCollationAndLimit(mq, join, join.getTraitSet().getCollation(),
+            null, null));
```

Comment:
Could you add a clarification why `true` is expected here?

Frankly speaking, it would be hard to maintain the test if it fails with "expected true got false" :-/

Reply:
`true` is expected because we are checking the `join` against its own collation (so of course it is already sorted by that collation) and no limit (fetch parameter is `null`). In any case, checking the result of the method is not really relevant, the original problem was a NPE inside `checkInputForCollationAndLimit`, so for this test we could perfectly remove the `assertTrue(...)`, and just the call `checkInputForCollationAndLimit`.

Reply:
It might be fine if you copy your explanation to the assert message: `() -> "we are checking the join against its own collation, fetch=null, offset=null => checkInputForCollationAndLimit must be true. join=" + join)`

WDYT?

>for this test I can perfectly remove the assertTrue(...), and just the call checkInputForCollationAndLimit.

If you remove all the asserts, then the test does not add much value. The code won't compile, and it would force the developer to handle null.

Reply:
The test without assert runs fine (I have just pushed a version in that way). It just executes a method that previously used to failed with a NPE, and now with the fix it runs successfully.
But I am ok with both (with or without assert). If you prefer, I can re-apply the assert with an explanation in the assert message.

Reply:
re-added assert, with message